### PR TITLE
APEXCORE-744 Add setting of predefined static logger appender properties

### PR DESCRIPTION
**Pull request by sgolovko**
Added setting of the static logger appender properties: application name, container id, user name, service name, node name.

@vrozov Please review the PR.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 1]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
+  private static String getService()
+  {
+    StackTraceElement[] stack = Thread.currentThread().getStackTrace();
+    String mainClass = stack[stack.length - 1].getClassName();
+    if (mainClass.equals("com.datatorrent.stram.cli.ApexCli")) {
+      return "apexclient";
+    } else if (mainClass.equals("com.datatorrent.stram.StreamingAppMaster")) {
+      return "apexmaster";
+    } else if (mainClass.equals("com.datatorrent.stram.engine.StreamingContainer")) {
+      return "apexcontainer";
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I would suggest to use apexworker instead. Also I would suggest to not use literals for above class names but use Class.getName() instead as those will need to be be changed to reflect org.apache.apex packages going forward.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 2]
```diff
@@ -62,6 +66,7 @@
 
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(LoggerUtil.class);
 
+  private static boolean setMDCProperties = true;
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Why is this necessary?

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 3]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
+        MDC.put("user", value);
+      }
+
+      value = System.getenv("CONTAINER_ID");
+      if (value != null) {
+        MDC.put("containerId", value);
+      }
+
+      Properties properties = System.getProperties();
+      String longName = LogicalPlan.APPLICATION_NAME.getLongName();
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Use DAG.Context instead of LogicalPlan

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 4]
```diff
@@ -375,6 +380,7 @@ public static boolean addAppender(String name, Properties properties)
       try {
         Method method = PropertyConfigurator.class.getDeclaredMethod("parseAppender", Properties.class, String.class);
         method.setAccessible(true);
+        makeMDCProperties();
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: MDC properties should be available for all appenders, including statically defined appenders

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 5]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
   }
+
+  /**
+   * Makes MDC properties
+   */
+  private static void makeMDCProperties()
+  {
+    if (setMDCProperties) {
+
+      MDC.put("node", getCurrentHost());
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Can we agree on the MDC properties names for apex. I'd suggest to use "org.apache.apex." or at minimum "apex." as a prefix for all MDC properties names.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 6]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
+      }
+
+      value = System.getenv("CONTAINER_ID");
+      if (value != null) {
+        MDC.put("containerId", value);
+      }
+
+      Properties properties = System.getProperties();
+      String longName = LogicalPlan.APPLICATION_NAME.getLongName();
+      if (properties.contains(longName)) {
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Avoid double get.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 7]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
+        MDC.put("application", properties.getProperty(longName));
+      }
+
+      if (MDC.get("service") == null) {
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Why is this necessary?

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 1 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Is the logic necessary at all? Does not the caller knows the service? Additionally, it will only work if called from the main thread. I'd prefer to avoid such dependency.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 8]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
...
+      return "apexcontainer";
+    }
+    return "unknown";
+  }
+
+  /**
+   * Returns the current host name
+   * @return The current host name
+   */
+  private static String getCurrentHost()
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Does not belong to this class. In case such helper method does not exist already, it should be added to one of the common utils classes.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 1 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: is it necessary to have "apex" in the name? Can it be "client", "master" and "worker"? apex is assumed.

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I agree with renaming of the service names and the usage the real class names instead of string constants.

But I'd prefer to have the current implicit logic for definition of the service name. If somebody needs to set the service name explicitly I can provide a method to do it.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 2 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: It is necessary in order to avoid the unnecessary reseting of the same static MDC properties for each appender.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 1 above)
Comment (tweise) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: +1 for removing apex from the name. Conciseness should be on our review checklist.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 4 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Ok. I've moved the initialization of the static appender properties to the static block. Also the method makeMDCProperties is public now. It will allow to call the initialization of the static appender properties explicitly (for instance, in unit tests). Also the new public method setServiceName(String service) is provided to set the service name explicitly.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 5 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: The long name of the MDC property can be inconvenient for mapping of the event property names with field/column names artifacts in table based storages (for instance, Elastic Search or any RDBMS).

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 3 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 6 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I don't see double get in the statement.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 7 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: We should not rewrite the  service name if it was set in the other code.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 2 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: It should be called only once and there is no need to call the function for each appender. In any case, MDC information is static and calling the method to populate MDC multiple times does not affect anything (MDC will be populated with the same information).

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 5 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: short name may collide with MDC properties populated by 3rd party libraries.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 6 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: contains() and get().

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 7 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: What is the use case where the function will be called with different names? Will this work for unit test and local stram mode (both app master and containers are in the same JVM, different threads)?

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 9]
```diff
@@ -464,4 +470,68 @@ public static boolean removeAppender(String name)
     }
     return names;
   }
+
+  /**
+   * Makes MDC properties
+   */
+  private static void makeMDCProperties()
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: consider renaming to setupMDC(String service)

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 6 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Ok.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 7 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: For instance, I'd like to share and reuse the method makeMDCProperties() in gateway code. If I call the method setServiceName() and provide the service name "gateway" I need to be sure that my service property will not be changed during implicit service update in the method  makeMDCProperties().

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: MDC properties are static in Java process. If both app master and containers are in the same JVM, both of them will have the same service name.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 9 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Ok

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 8 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

**Comment by sgolovko** [comment]
@tweise @vrozov I've updated the PR code according to your comments. Please look at the updates.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 10]
```diff
@@ -73,6 +79,10 @@ public String apply(@Nullable Level input)
     }
   };
 
+  static {
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I prefer to avoid static initialization. It will not work for unit tests where both app master and containers/workers run in the same JVM but different threads.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 11]
```diff
@@ -464,4 +474,55 @@ public static boolean removeAppender(String name)
...
     return names;
   }
+
+  /**
+   * Makes MDC properties
+   */
+  public static void setupMDC()
+  {
+    String node = StramClientUtils.getHostName();
+    MDC.put("node", node == null ? "unknown" : node);
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Instead of "unknown", use Sting representation of IP address.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 12]
```diff
@@ -464,4 +474,55 @@ public static boolean removeAppender(String name)
...
+    if (value != null) {
+      MDC.put("user", value);
+    }
+
+    value = System.getenv("CONTAINER_ID");
+    if (value != null) {
+      MDC.put("containerId", value);
+    }
+
+    value = System.getProperties().getProperty(APPLICATION_NAME.getLongName());
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: System.getProperty()?

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 13]
```diff
@@ -464,4 +474,55 @@ public static boolean removeAppender(String name)
...
+    if (MDC.get("service") == null) {
+      MDC.put("service", getService());
+    }
+  }
+
+  /**
+   * Returns the service name
+   * @return Service name
+   */
+  private static String getService()
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I am not convinced that getting service name implicitly is required. There are just two or three places where an explicit call to setupMDC() will be added, in case it is explicit and it will help in case later somebody needs to understand where and how service names are set. The implicit setup makes it much harder to track. I would agree with the implicit initialization in case it would be necessary to insert setupMDC() in multiple locations.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 14]
```diff
@@ -34,13 +34,13 @@
 import javax.annotation.Nullable;
 
 import org.apache.apex.log.LogFileInformation;
-
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Avoid format changes.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 10 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: The usage of the static block is a general approach in Java development for initialization static properties. I think it is the right way.

MDC properties are static properties. So if both app master and workers are running in the same JVM, MDC properties will have the same values in all threads.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 11 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: The method java.net.getHostName() itself returns ip-address, if it is not able to retrieve the host name.

The implementation of our method getHostName() is:

  public static String getHostName()
  {
    try {
      return java.net.InetAddress.getLocalHost().getHostName();
    } catch (UnknownHostException ex) {
      return null;
    }
  }

The method java.net.InetAddress.getLocalHost() can throw the exception UnknownHostException if the local host could not be resolved into an address.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 12 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Agree. It is the same. Done.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 13 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: I'm sure the implicit setting of the MDC properties with the default values is much convenient and the right approach. If some caller needs to setup the MDC properties explicitly, the caller can call the methods setupMDC() and setServiceName()  explicitly.

And I don't think "the implicit setup makes it much harder to track" is the right statement. It is absolutely generic approach in the Java development to set the default properties values.

I think 3 locations (client, master, worker) are "multiple locations".

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 14 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

**Comment by sgolovko** [comment]
@vrozov  @tweise The PR is updated with the minor changes according to the review comments.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 10 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: MDC properties are not static. Each thread may have it's own MDC properties populated.

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: MDC properties are static.

public class MDC {
  
  final static MDC mdc = new MDC();

  static  public  void put(String key, Object o) {
     if (mdc != null) {
         mdc.put0(key, o);
     }
  }

Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Please check log4j java doc: "The MDC is managed on a per *thread* basis. A child thread automatically inherits a copy of the mapped diagnostic context of its parent."

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 13 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: It is not generic as it will not work unless the method is invoked from the main thread. If another developer modifies "main" class for the master or workers and introduces another "main" class, the MDC functionality will be broken too.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 10 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Agree. The implementation of MDC allows to set MDC properties for a thread.

But I disagree with your statement "It will not work for unit tests where both app master and containers/workers run in the same JVM but different threads." If somebody needs to implement the unit test that would run the master and containers in different threads, he or she can call the method setService() in the unit test code explicitly.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 13 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Unfortunately it is impossible to prevent all possible stupid mistakes. We can only try to minimize the number of mistakes. And if we are going to make it the setting of MDC properties MUST be done in the thread with the main method in the root of the program stack only because the thread of the main class is a parent of all other threads.

I've added the check that the implicit MDC initialization is called from the main thread.

    StackTraceElement[] stack = Thread.currentThread().getStackTrace();
    if (!"main".equals(stack[stack.length - 1].getMethodName())) {
      throw new RuntimeException("The implicit initialization of MDC properties must be done from the main thread only");
    }

But if another developer modifies "main" class for the master or workers and introduces another "main" class, it will be easy checked and fixed, because the service name in logs will have the value "unknown".

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Replaced exception with assert.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 10 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: The approach to use different threads for master and containers/workers is already implemented in the StramLocalCluster. Running applications in the local mode and in unit tests that use StramLocalCluster should not require any additional explicit calls to the setService().

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 13 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: The entire function (15 lines of code) can be replaced with 3 explicit calls to setupMDC().

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: ok. I've added explicit initialization  of MDC properties.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Apex_Core_PR/851/

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 15]
```diff
@@ -464,4 +472,33 @@ public static boolean removeAppender(String name)
     }
     return names;
   }
+
+  /**
+   * Makes MDC properties
+   */
+  public static void setupMDC(String service)
+  {
+    MDC.put("dt.service", service);
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Please use "apex" prefix.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 16]
```diff
@@ -464,4 +472,33 @@ public static boolean removeAppender(String name)
...
+
+    String value = StramClientUtils.getHostName();
+    MDC.put("dt.node", value == null ? "unknown" : value);
+
+    value = System.getenv(ApplicationConstants.Environment.USER.name());
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: use USER.key() or USER.toString()

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java [hunk 17]
```diff
@@ -35,12 +35,17 @@
 
 import org.apache.apex.log.LogFileInformation;
 
+import org.apache.hadoop.yarn.api.ApplicationConstants;
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: import org.apache.hadoop.yarn.api.ApplicationConstants.Environment

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 18]
```diff
@@ -4129,7 +4129,8 @@ public void mainHelper() throws Exception
 
   public static void main(final String[] args) throws Exception
   {
-    LoggerUtil.addAppenders();
+//    LoggerUtil.setupMDC("client");
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Why is the change necessary? If it is required, do not use comments, delete lines.

Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 17 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 16 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java (hunk 15 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java (hunk 18 above)
Comment (sgolovko) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Added a call to init MDC properties.

**Comment by vrozov** [comment]
Please fix unit test

**Comment by sgolovko** [comment]
Fixed the unit test.

**Comment by vrozov** [comment]
LGTM. Please rebase against upstream/master

**Comment by sgolovko** [comment]
Done.
